Account for multiple project roots

PEZ opened the pull request:
Glob for all project roots in the workspace

Let the user pick if there are more than one
Changes the way we find the closest project root
Not using node fs module at all
Fixes #1254

## My Calva PR Checklist
<!-- Strike out (using `~`) items that do not apply, as Github reports how many are not ticked. If you want to add checkboxes, please do. -->

I have:

- [x] Read [How to Contribute](https://github.com/BetterThanTomorrow/calva/wiki/How-to-Contribute#before-sending-pull-requests).
- [x] Directed this pull request at the `dev` branch. (Or have specific reasons to target some other branch.)
- [x] Made sure I have changed the PR base branch, so that it is not `published`. (Sorry for the nagging.)
- [x] Updated the `[Unreleased]` entry in `CHANGELOG.md`, linking the issue(s) that the PR is addressing.
- [x] Figured if **anything** about the fix warrants tests on Mac/Linux/Windows/Remote/Whatever, and either tested it there if so, or mentioned it in the PR.
- [x] Added to or updated docs in this branch, if appropriate
- [x] Tests
     - [x] Tested the particular change
     - [x] Figured if the change might have some side effects and tested those as well.
     - [x] Smoke tested the extension as such.
     - [x] Tested the VSIX built from the PR (so, after you've submitted the PR). You'll find the artifacts by clicking _Show all checks_ in the CI section of the PR page, and then _Details_ on the `ci/circleci: build` test.
- [x] Referenced the issue I am fixing/addressing _in a commit message for the pull request_.
     - [x] If I am fixing the issue, I have used [GitHub's fixes/closes syntax](https://help.github.com/en/articles/closing-issues-using-keywords)
- [x] Created the issue I am fixing/addressing, if it was not present.
- [x] Formatted all JavaScript and TypeScript code that was changed. (use the [prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode) or run `npm run prettier-format`)
- [x] Confirmed that there are no linter warnings or errors (use the [eslint extension](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint), run `npm run eslint` before creating your PR, or run `npm run eslint-watch` to eslint as you go).

<!-- This is a nice book to read about the power of checklists: https://www.samuelthomasdavies.com/book-summaries/health-fitness/the-checklist-manifesto/ -->

Ping @pez, @bpringe

PEZ committed:
Glob for all project roots in the workspace

Let the user pick if there are more than one
Changes the way we find the closest project root
Not using node fs module at all
Fixes #1254

PEZ committed:
prettier

corasaurus-hex commented on src/state.ts:
```diff
@@ -134,100 +135,66 @@ function getProjectWsFolder(): vscode.WorkspaceFolder | undefined {
+async function findProjectRootPaths() {
+  const projectFileNames: string[] = ['project.clj', 'shadow-cljs.edn', 'deps.edn'];
+  const projectFilesGlob = `**/{${projectFileNames.join(',')}}`;
+  const excludeDirsGlob = `**/{${config.getConfig().projectRootsSearchExclude.join(',')}}`;
```
I see `target` is in the list of excluded directories. Would that exclude a directory named `target-files` or does it need to be just `target` by itself in order to match this glob?

PEZ replied:
It will only exclude `target`, exactly as that. Wherever it appears, though...

corasaurus-hex commented on src/state.ts:
```diff
@@ -134,100 +135,66 @@ function getProjectWsFolder(): vscode.WorkspaceFolder | undefined {
+  console.debug('glob took', new Date().getTime() - t0, 'ms');
+  const projectFilePaths = candidateUris.map((uri) => path.dirname(uri.fsPath));
+  const candidatePaths = [...new Set(projectFilePaths)];
+  console.log(candidatePaths);
```
It might be cool to make this `console.log({candidatePaths})` so the log entry contents are labeled?

PEZ replied:
TIL.

Wasn't intending to keep this logging in there, though. Except the timing of the glob, which could be good to be able to ask users about for a while.

corasaurus-hex commented on src/state.ts:
```diff
@@ -134,100 +135,66 @@ function getProjectWsFolder(): vscode.WorkspaceFolder | undefined {
+    ? candidatePaths
+        .filter((p) => docDir.startsWith(p))
+        .sort()
+        .reverse()[0]
```
Nice! That's a good way to find the longest match 😄

corasaurus-hex reviewed (approved):
I had a couple questions but other than that this is beautiful 😚👌🏻

corasaurus-hex commented on src/config.ts:
```diff
@@ -161,6 +161,7 @@ function getConfig() {
       useDeprecatedAliasFlag: configOptions.get<boolean>('jackIn.useDeprecatedAliasFlag'),
     },
     enableClojureLspOnStart: configOptions.get<boolean>('enableClojureLspOnStart'),
+    projectRootsSearchExclude: configOptions.get<string[]>('projectRootsSearchExclude'),
```
Oh! I missed one thing! We should probably add a default here so that it satisfies the gods of `strictNullChecks`:

```
configOptions.get<string[]>('projectRootsSearchExclude') ?? []
```

PEZ replied:
Thanks. I used the default value signature of the `get`:

```ts
configOptions.get<string[]>('projectRootsSearchExclude', []) 
```

Does that also satisfy the rule? Less syntax to parse...

corasaurus-hex replied:
ooooh, I didn't know about that one! I think that should fix it? and If not I can fix it later when i'm fixing null check things 😊

PEZ committed:
Update docs about monorepo support

PEZ committed:
Update connect to avoid two root menus

And generallt just clean up the logic some

PEZ committed:
prettier

PEZ committed:
Add title option to quickPickSingle

PEZ committed:
Sort the project roots alphabetically

PEZ committed:
Provide testDirUri to initProjectDir to make e2e test avoid menu

PEZ committed:
Experimental quickPick selection for e2e tests

PEZ commented:
@Cyrik : can you have a look at e356cfd and see if you think it is viable way to deal with the quick-pick menus in e2e tests? I didn't figure out a very nice way to select the correct sequence...

I'd be awesome if @corasaurus-hex and @bpringe had a look as well.

:heart:

PEZ commented:
I wonder why that `saveAs` doesn't do the trick now... I'll have a closer look. Thanks!

corasaurus-hex commented on src/connector.ts:
```diff
@@ -668,27 +659,29 @@ async function standaloneConnect(
+    let projectDirUri = state.getProjectRootUri();
+    if (!projectDirUri) {
+      projectDirUri = await state.getOrCreateNonProjectRoot(context, true);
+    }
```
it looks like `getOrCreateNonProjectRoot` calls `getProjectRootUri` internally, first thing, and returns it later. Could we ditch the `getProjectRootUri` here and just call `getOrCreateNonProjectRoot` on its own here? Or should we remove the `getProjectRootUri` from `getOrCreateNonProjectRoot`? Or just leave everything as-is?

PEZ replied:
I'll have to check and think. You are probably right. This has been somewhat cleaned up, but I might have missed this opportunity. Tomorrow, though, because right now too 💤 !

corasaurus-hex commented on src/connector.ts:
```diff
@@ -668,27 +659,29 @@ async function standaloneConnect(
+    if (!projectDirUri) {
+      projectDirUri = await state.getOrCreateNonProjectRoot(context, true);
+    }
+    await state.initProjectDir(projectDirUri);
```
We add a `.catch` to this function in `connectCommand`, do we need one here?

PEZ replied:
Iirc i added the one in `connectCommand` because the user can do funny stuff with the menu. Here we don't allow that. But it could also go wrong because of non-validated glob patterns, so yes, we should have one here anyway. =)

corasaurus-hex commented on src/utilities.ts:
```diff
@@ -547,6 +558,7 @@ export {
   logSuccess,
   getCljsReplStartCode,
   getShadowCljsReplStartCode,
+  quickPickActive as quicPickActive,
```
Can we get rid of the `as quicPickActive` here? We'll need to fix the references to it in the tests, too.

PEZ replied:
No idea why this coercion is here... I'll ditch it!

corasaurus-hex reviewed (approved):
I think this looks great! I love to see code improvements like this.

PEZ committed:
Move project root stuff to project-root module

PEZ committed:
Less stateful way to get project root

PEZ committed:
prettier

PEZ commented:
> I love to see code improvements like this.

❤️ 

When I tried to extract the find-root-paths stuff to a module named `project-root.ts`, it turned out there was already such a module! And it had yet another implementation of the same traversing! Crazy. I just replaced it with the new stuff and updated the only call site. It seems to work. But also makes us glob more often, of course. Maybe we should keep the other way to find closest root around anyway...

bpringe commented on docs/site/connect.md:
```diff
@@ -61,6 +63,19 @@ All this said, I still recommend you challenge the conclusion that you can't use
+
+![The project roots menu](images/calva-monorepo-project-roots-menu.png)
+
+When searching for project roots in your workspace, Calva will glob for all files matching `project.clj`, `deps.edn`, or `shadow-cljs.edn`. This is done using VS Code's workspace search engine, and is very efficient. However, in a large monorepo, it is still a sufficient task. In order to not waste resources Calva will exclude any directories in the setting `calva.projectRootsSearchExclude`. 
```
Did you mean to say "still a substantial task" here? It seems, judging by the context, that you mean to point out that it even though the search engine is very efficient, a large monorepo might still take a noticeable amount of time.

PEZ replied:
Haha, yes. Thanks!

bpringe commented on src/project-root.ts:
```diff
@@ -1,54 +1,50 @@
+          title: 'Project root',
+          values: candidatePaths,
+          default: closestRootPath,
+          placeHolder: 'Multiple Clojure projects found, pick one',
```
```suggestion
          placeHolder: 'Multiple Clojure projects found; pick one.',
```

PEZ committed:
Update src/project-root.ts

Co-authored-by: Brandon Ringe <brandon.ringe@gmail.com>

PEZ committed:
Fix typo

PEZ committed:
Typo

PEZ committed:
Clean up some more statefulness

PEZ committed:
Merge branch '1254-multiple-project-roots' of github.com:BetterThanTomorrow/calva into 1254-multiple-project-roots

PEZ committed:
Remove some debug logging

PEZ commented:
🎉
